### ISSUE TITLE
Fallback to Composite Implicit Autograd for Purely Functional Composite Custom Ops

### ISSUE THREAD

**Issue by asglover** [pytorch/pytorch#174165]
### 🚀 The feature, motivation and pitch

I'm a maintainer for a project [OpenEquivariance](https://github.com/PASSIONLab/OpenEquivariance) which provides specialized CUDA/HIP kernels for scientific ML applications. We use CustomOp to do this right now but there is one major shortcoming, which if addressed would make life much easier. 

The challenge is that we have a CustomOp definition which describes a forward pass. It is a pure composite op, in that it can be described exactly as sequence of PyTorch native ops. It's not necessary to know what the op looks like, but if you're curious, it looks something like this: [link](https://github.com/e3nn/e3nn/blob/ada35561ee184e710b0f6430d2cbe45cb6e79075/e3nn/o3/experimental/_full_tp.py#L65)

The challenge is that we register the CustomOp and then distribute the project to scientists who continuously want to take higher and higher derivatives backwards through the custom op. We do actually provide several fast backwards passes, but the number of derivatives is high, and in order to support new hardware, we are obligated to implement all of the levels of backward passes currently in use.  

It would be greatly appreciated if something like the "fallback" ability which is currently restricted to "_" namespace operators could be extended to custom ops which are pure composite functions. So that if a backwards kernel is not defined, then it can be derived from the `CompositeImplicitAutograd` kernel. Ideally I would want the forward (and any specifically defined backwards passes) to go through their specialized kernels, and any undefined higher order backwards passes to through the composite path. 

This would allow framework managers to ditch the special dispatching logic and users to confidently be able to experiment with higher derivatives without losing the performance of the specialized kernels. 

I have a lot of flexibility in how I register the CustomOps and am happy to use lower level APIs to achieve this. 
If you know what work needs to happen in PyTorch to make this behavior possible, I would be happy to work towards a PR to make it happen. 

@zou3519 @ezyang 

### Alternatives

To work around this currently the various frameworks that use which use our kernels implement extensive logic in the model builders to manually substitute our CustomOp for the PyTorch-native implementation. End users control this with a .yaml file and a model builder. This has the additional downside of forcing users to not benefit fast forward, backward, double backward, all because they need to additionally do one triple backward. It would be much cleaner if falling back to the CompositeAutograd implementation for unsupported derivatives was handled by pytorch. 

I also looked in to re-dispatching to a PyTorch native implementation. This helps in some cases, but for this specific task, falling back to the pytorch native implementation for unspecified backwards passes, I haven't been able to make it work exactly how I want it. I created a demo repo where I explore these ideas. 

I made sure that this is very readable and reproducible, you should be able to get the same results without trouble. 

[sandbox](https://github.com/asglover/testing-pytorch-extensions)

There are test cases which try registering the extension with various combinations of C++ and python. The C++ based dispatcher will re-dispatch to the CompositeAutograd solution but it does it re-dispatches to the composite autograd pass for both the forward and backward. I would really like the specialized C++ kernel to be used for forwards and the composite autograd for backward. 

Maybe a custom dispatch class is what is needed? I'm not sure. 

If you feel like there is a way to achieve this goal in PyTorch today, please let me know. 

### Additional context

The concept of having a "CustomOp" be defined by a composite of PyTorch native ops which can also be used as a fall back implementation is a core part of the VLLM proposal for their VLLM IR. I independently tried to solve achieve this with existing pytorch tools, and was amused to read the proposal. I know there is a goal to upstream aspects of VLLM's work. I think this would be a great example of a feature which would help the broader community. I'm sure there are other users that would benefit from this feature. 

[vllm ir proposal](https://docs.google.com/document/d/1zyPdGVX7TnVtvqD9NsaTO4E02u7PY7uudreLrdWp2GM/edit?tab=t.xerdhb9i406r)

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan @bobrenjc93 @chauhang @penguinwu @bdhirsh @aorenste

**Comment by asglover** [comment]
Hello Richard, 

> I can't see the "[sandbox](https://github.com/asglover/testing-pytorch-extensions)" link.

Sorry that was a mistake, it should be public now. 

> I assume you have multiple autograd.Function

I was not directly using autograd.Function because the documentation warned against directly handling them, but I'm happy to directly use autograd.Function if that is the way. 

> you can check is_grad_enabled() in the backward of the autograd.Function. If it is enabled, and you do not have a further backward implementation, then you should call the decomposed version of the operation. Otherwise, if grad is disabled (this means that the user doesn't need additional gradients), then you can call the fast version.

In my test case, example, where I'm only calling one derivative, and autograd is on, I end up having to dispatch to the slow decomposed version. Is this a fundamental pytorch thing? Is there no way to get the fast implementation for the forwards and autograd backwards through the composite? 

I think I'm doing what you describe in the function registered to `AutogradCPU`. 

```
#include <torch/extension.h>
#include <torch/library.h>

constexpr int NUM = {{NUM}}; 

torch::Tensor num_add_cpu(
    const torch::Tensor& a,
    int64_t b) {
  TORCH_CHECK(a.scalar_type() == torch::kFloat, "expected float tensor");
  TORCH_CHECK(a.device().is_cpu(), "expected CPU tensor");
  TORCH_CHECK(b == NUM, "expected num to match build-time NUM");

  torch::Tensor a_contig = a.contiguous();
  torch::Tensor result = torch::empty_like(a_contig);

  const float* a_ptr = a_contig.const_data_ptr<float>();
  float* result_ptr = result.mutable_data_ptr<float>();
  std::cout << "USING SPECIALIZED C++ EXTENSION" << std::endl; 
  for (int64_t i = 0; i < result.numel(); i++) {
    result_ptr[i] = a_ptr[i] + NUM;
  }
  return result;
}

torch::Tensor forward_op_autograd(const torch::Tensor& a, int64_t b) {
  static auto op =
      c10::Dispatcher::singleton()
          .findSchemaOrThrow("{{LIBRARY_NAME}}::forward_op", "")
          .typed<torch::Tensor(const torch::Tensor&, int64_t)>();
  
  std::cout << "C++ REDISPATCH" << std::endl; 

  const bool grad_enabled = at::GradMode::is_enabled();
  const bool a_needs_grad = a.requires_grad();

  if (grad_enabled && a_needs_grad){
    return op.redispatch(c10::DispatchKeySet(c10::DispatchKey::CompositeImplicitAutograd), a, b);
  } else {
    return op.redispatch(c10::DispatchKeySet(c10::DispatchKey::CPU), a, b);
  }
}


TORCH_LIBRARY_IMPL({{LIBRARY_NAME}}, CPU, m) {
  m.impl("forward_op", num_add_cpu);
};

TORCH_LIBRARY_IMPL({{LIBRARY_NAME}}, AutogradCPU, m) {
  m.impl("forward_op", forward_op_autograd);
}

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {}
```

This tests case where I call backwards: 

```
def test_backward(module, t):
    t_test = t.clone().requires_grad_()
    t_ref = t.clone().requires_grad_()

    with DebugMode(record_stack_trace=True) as dm:
        loss_test = torch.sum(module(t_test))
        loss_test.backward()

    loss_ref = torch.sum(t_ref + module.number)

    loss_ref.backward()
    logger.debug(f"{t_test.grad}")
    logger.debug(f"{t_ref.grad}")

    logger.debug(dm.debug_string())
    torch.testing.assert_close(t_test.grad, t_ref.grad)
```

And it works but it never goes through my fast forward implementation
